File: GUI/WpfMidiOrganPedals/Devices/RawMessagePacker.cs
using System;

namespace WpfMidiOrganPedals.Devices
{
    internal class RawMessagePacker : RawMessageHandler
    {
        internal byte[] Pack(byte id, byte[] payloadData)
        {
            var result = new byte[4 + payloadData.Length];

            result[0] = StartByte;
            result[1] = id;
            result[2] = (byte)payloadData.Length;
            result[3] = CalcChecksum(payloadData);
            Array.Copy(payloadData, 0, result, 4, payloadData.Length);

            return result;
        }
    }
}

File: Firmware/RawMessagePacker.cpp
#include "RawMessagePacker.h"
#include "ChecksumCalculator.h"


void RawMessagePacker::Pack(const RawMessage &input)
{
  uint8_t id = input.GetId();
  const uint8_t *payloadData = input.GetData();
  uint8_t payloadSize = input.GetSize();
  uint8_t checksum = ChecksumCalculator::CalcChecksum(payloadData, payloadSize);
  
  data[0] = StartByte;
  data[1] = id;
  data[2] = payloadSize;
  data[3] = checksum;
  memcpy(data+4, payloadData, payloadSize);
  size = 4 + payloadSize;
}


uint16_t RawMessagePacker::GetSize() const
{
  return size;
}


const uint8_t *RawMessagePacker::GetData() const
{
  return data;
}

